flang/lib/Lower/Allocatable.cpp

```diff
@@ -1198,6 +1231,75 @@ void Fortran::lower::disassociateMutableBox(
   MutablePropertyWriter{builder, loc, box}.setUnallocatedStatus();
 }
 
+void Fortran::lower::genReallocIfNeeded(Fortran::lower::FirOpBuilder &builder,
+                                        mlir::Location loc,
+                                        const fir::MutableBoxValue &box,
+                                        mlir::ValueRange lbounds,
+                                        mlir::ValueRange shape,
+                                        mlir::ValueRange lengthParams) {
+  // Implement 10.2.1.3 point 3 logic when lhs is an array.
+  auto reader = MutablePropertyReader(builder, loc, box);
+  auto addr = reader.readBaseAddress();
+  auto isAllocated = genIsNonNull(builder, loc, addr);
+  auto ifAllocated = builder.create<fir::IfOp>(
+      loc, mlir::TypeRange{}, isAllocated, /*withElseRegion=*/true);
+  builder.setInsertionPointToStart(&ifAllocated.thenRegion().front());
+  { // If the box is allocated check if it must be reallocated and reallocate.
```

Comment:
Can you go ahead and name this code but provide a stub?

```c++
static bool isBoxAllocatedButNeedsReallocation(...) { return true; }

...
  if (isBoxAllocatedButNeedsReallocation(foo)) {
```

I'm not sure how that will play here as it looks like both the `if` and `else` arms execute. Hmm.

Reply:
Yes, I was missing a bit of structure and hence used the `{}`  code blocks. I went further and just added a FirOpBuilder helper to generate if-then-else like:
```
builder.genIfThenElse(loc, cdt)
    .genThen([&]() { 
      /* generate then code */ 
     })
    .genElse([&](){
      /* generate else code */
     })
    .end();
```

Without having to deal with the insertion points. The only thing I do not like is that I feel the four spaces of clang-format are a bit too much indentation when piping calls. But it is what it is. Tell me if you think it is overkill, I'll just drop that change.

Reply:
That's easier to read (IMHO). Let's go with it. We might want to use the pattern in some other places to see how it looks there as well. It does abstract away all the insertion point twiddling, so that's an improvement.